### PR TITLE
Perf and stability changes for load

### PR DESCRIPTION
**Pull request by tomlm**
- fix autofac leak
- change InMemoryStore to be MemoryCache based
- Change CachedMemoryStore to be MemoryCache based
- Fixed 401 problem with idle bots
- Remove gzip compression from InMemoryStore
- removed global HttpClient singleton hack
- Changed ConnectorClientFactory and autofac registration to have singleton for ConnectorClient and StateClient

### CODE REVIEW

File: CSharp/Library/Microsoft.Bot.Builder/ConnectorEx/BotData.cs [hunk 1]
```diff
@@ -89,18 +91,25 @@ public interface IBotDataStore<T>
     /// </summary>
     public class InMemoryDataStore : IBotDataStore<BotData>
     {
-        internal readonly ConcurrentDictionary<string, string> store = new ConcurrentDictionary<string, string>();
+        internal static readonly MemoryCache store = new MemoryCache(nameof(InMemoryDataStore), new NameValueCollection() { { "PhysicalMemoryLimitPercentage", "50" } });
```
Comment (willportnoy) [inline CSharp/Library/Microsoft.Bot.Builder/ConnectorEx/BotData.cs]: >static [](start = 17, length = 6)

Doesn't InMemoryDataStore have lifetime specified in the container?  Making it static interferes with testability.

File: CSharp/Library/Microsoft.Bot.Builder/ConnectorEx/BotData.cs [hunk 2]
```diff
@@ -271,7 +262,8 @@ public enum CachingBotDataStoreConsistencyPolicy
     public class CachingBotDataStore : IBotDataStore<BotData>
     {
         private readonly IBotDataStore<BotData> inner;
-        internal readonly Dictionary<IAddress, CacheEntry> cache = new Dictionary<IAddress, CacheEntry>();
+        internal static readonly MemoryCache cache = new MemoryCache(nameof(CachingBotDataStore), new NameValueCollection() { { "PhysicalMemoryLimitPercentage", "50" } });
```
Comment (willportnoy) [inline CSharp/Library/Microsoft.Bot.Builder/ConnectorEx/BotData.cs]: >static [](start = 17, length = 6)

CachingBotDataStore probably has a lifetime specified in the container.

File: CSharp/Library/Microsoft.Bot.Builder/ConnectorEx/BotData.cs [hunk 3]
```diff
@@ -280,23 +272,31 @@ public CachingBotDataStore(IBotDataStore<BotData> inner, CachingBotDataStoreCons
...
         internal class CacheEntry
         {
             public BotData BotConversationData { set; get; }
             public BotData BotPrivateConversationData { set; get; }
             public BotData BotUserData { set; get; }
         }
 
+        private string GetCacheKey(IAddress key)
+        {
+            return $"{key.BotId}.{key.BotId}.{key.ConversationId}.{key.UserId}";
```
Comment (willportnoy) [inline CSharp/Library/Microsoft.Bot.Builder/ConnectorEx/BotData.cs]: >{key.BotId}.{key.BotId} [](start = 21, length = 23)

why repeat?

File: CSharp/Library/Microsoft.Bot.Builder/ConnectorEx/IConnectorClientFactory.cs [hunk 4]
```diff
@@ -60,12 +62,45 @@ public sealed class ConnectorClientFactory : IConnectorClientFactory
         private readonly Uri serviceUri;
         private readonly IAddress address;
         private readonly MicrosoftAppCredentials credentials;
+        private readonly ConnectorClient connectorClient;
+        private readonly StateClient stateClient;
+
+        private static readonly ConcurrentDictionary<string, ConnectorClient> connectorClients = new ConcurrentDictionary<string, ConnectorClient>();
```
Comment (willportnoy) [inline CSharp/Library/Microsoft.Bot.Builder/ConnectorEx/IConnectorClientFactory.cs]: >static [](start = 16, length = 6)

interferes with lifetime specified in container

File: CSharp/Library/Microsoft.Bot.Builder.Autofac/Fibers/DoNotSerializeResolver.cs [hunk 5]
```diff
@@ -97,20 +97,42 @@ bool IResolver.TryResolve(Type type, object tag, out object value)
                     IComponentRegistration registration;
                     if (registry.TryGetRegistration(service, out registration))
                     {
-                        // Autofac will still generate "implicit relationship types" (e.g. Func or IEnumerable)
-                        // and ignore the key in KeyedService
-                        bool generated = registry.IsRegistered(new KeyedService(new object(), serviceType));
-                        if (!generated)
+                        if (IsAutoFacImplicit(serviceType))
```
Comment (willportnoy) [inline CSharp/Library/Microsoft.Bot.Builder.Autofac/Fibers/DoNotSerializeResolver.cs]: >IsAutoFacImplicit [](start = 28, length = 17)

I thought we pushed this filter into the Services cache above?

File: CSharp/Library/Microsoft.Bot.Connector.Shared/MicrosoftAppCredentials.cs [hunk 6]
```diff
@@ -175,53 +175,31 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
...
+                    // check to see if our token is expired 
+                    if (TokenExpired(oAuthToken))
                     {
-                        // if we are past the refresh time
-                        if (DateTime.UtcNow > refreshTime)
+                        // it is, we should await the current task (someone could have already asked for a new token)
+                        oAuthTokenTask = _getCurrentTokenTask(forceRefresh:false);
+
+                        // if the task is completed and is the expired token, then we need to force a new one 
+                        // (This happens if bot has been 100% idle for an past the expiration point)
```
Comment (willportnoy) [inline CSharp/Library/Microsoft.Bot.Connector.Shared/MicrosoftAppCredentials.cs]: >for an past [](start = 67, length = 11)

missed a word

File: CSharp/Library/Microsoft.Bot.Builder/Microsoft.Bot.Builder.csproj [hunk 7]
```diff
@@ -57,6 +57,7 @@
     <Reference Include="System.Core" />
     <Reference Include="System.Net" />
     <Reference Include="System.Net.Http.WebRequest" />
+    <Reference Include="System.Runtime.Caching" />
```
Comment (willportnoy) [inline CSharp/Library/Microsoft.Bot.Builder/Microsoft.Bot.Builder.csproj]: >System.Runtime.Caching [](start = 24, length = 22)

this is available in .Net standard?

Comment (tomlm) [reply CSharp/Library/Microsoft.Bot.Builder/Microsoft.Bot.Builder.csproj]: Yes, [https://github.com/aspnet/Caching/tree/dev/src/Microsoft.Extensions.Caching.Memory](https://github.com/aspnet/Caching/tree/dev/src/Microsoft.Extensions.Caching.Memory) 

---
In reply to: [175547434](https://github.com/Microsoft/BotBuilder/pull/4332#discussion_r175547434) [](ancestors = 175547434)

**Pull request merged**